modules/users/server/controllers/users/users.profile.server.controller.js

```diff
@@ -60,39 +60,83 @@ exports.changeProfilePicture = function (req, res) {
   var user = req.user;
   var upload = multer(config.uploads.profileUpload).single('newProfilePicture');
   var profileUploadFileFilter = require(path.resolve('./config/lib/multer')).profileUploadFileFilter;
+  var tmpImageUrl;
 
   // Filtering to upload only images
   upload.fileFilter = profileUploadFileFilter;
-
+  
   if (user) {
-    upload(req, res, function (uploadError) {
-      if (uploadError) {
-        return res.status(400).send({
-          message: 'Error occurred while uploading profile picture'
-        });
-      } else {
-        user.profileImageURL = config.uploads.profileUpload.dest + req.file.filename;
+    tmpImageUrl = user.profileImageURL;
+    uploadImage()
+      .then(updateUser)
+      .then(deleteOldImage)
+      .then(login)
+      .then(function () {
+        res.json(user);
+      })
+      .catch(function (err) {
+        res.status(400).send(err);
+      });
+  } else {
+    res.status(400).send({
+      message: 'User is not signed in'
+    });
+  }
+
+  function uploadImage () {
+    return new Promise(function (resolve, reject) {
+      upload(req, res, function (uploadError) {
+        if (uploadError) {
+          reject({
+            message: 'Error occurred while uploading profile picture'
```

Comment:
Thank you @simison! This looks very handy.

@aanev Could something like this be incorporated into our [existing error handling](https://github.com/meanjs/mean/blob/master/modules/core/server/controllers/errors.server.controller.js) server-side module?

These codes could be useful for the client application.